docs/_posts/2015-02-24-streamlining-react-elements.md

```diff
@@ -0,0 +1,230 @@
+
+### Solution: Make Context Parent-Based Instead of Owner-Based
+
+The first thing we’re doing is warning you if you’re using the "owner" feature in a way that relies on it propagating through owners. Instead, we’re planning on propagating it through parents to its children. In almost all cases, this shouldn’t matter. In fact, parent-based contexts is simply a superset.
+
+### Solution: Remove the Semantic Implications of Owner
+
+It turns out that the cases where owners as part of state-semantics. As a precaution, we’ll warn you if it turns out that the owner is important to determine state. In almost every case this shouldn’t matter. Unless you’re doing some weird optimizations, you shouldn’t see this warning.
+
+### Pending: Change the refs Semantics
+
+Refs are still based on "owner". We haven’t fully solved this special case just yet.
+
+In 0.13 we introduced a new callback-refs API that doesn’t suffer from these problems but we’ll keep on a nice declarative alternative to the current semantics for refs. As always, we won’t deprecate something until we’re sure that you’ll have a nice upgrade path.
+
+## Keyed Objects as Maps
+
+In React 0.12 you could use keyed objects to provide an external key to an element or a set. This pattern isn’t actually widely known, nor documented outside of Facebook. It shouldn’t be an issue for most of you.
+
+```js
+<div>{{ a: <span />, b: <span /> }}</div>
+```
+
+### Problem: Relies on Enumeration Order
+
+The problem with this pattern is that it relies on enumeration order of objects. This is technically unspecified, even though implementations now agree to use insertion order. Except for the special case when numeric keys are used.
+
+### Problem: Using Objects as Maps is Bad
+
+It is generally accepted that using objects as maps screw up type systems, VM optimizations, compilers etc. It is my better to use a dedicated data structure like ES6 Maps.
+
+More importantly, this can have important security implications. For example this has a potential security problem:
+
+```js
+var children = {};
+items.forEach(item => x[item.title] = <span />);
+return <div>{children}</div>;
+```
+
+Imagine if `item.title === '__proto__'` for example.
+
+### Problem: Can’t be Differentiated from Arbitrary Objects
+
+Since these objects can have any keys with almost any value, we can’t differentiate them from a mistake. If you put some random object, we will try our best to traverse it and render it, instead of failing with a helpful warning. In fact, this is one of the few places where you can accidentally get an infinite loop in React.
+
+To differentiate ReactElements from one of these objects, we have to tag them with `_isReactElement` which we wouldn’t need to without this.
```

Comment:
> To differentiate ReactElements from one of these objects, we have to tag them with `_isReactElement` ~~which we wouldn’t need to without this~~.

Drop that 2nd part. If it feels too short now you can also qualify with another sentence. "This is another issue preventing us from inlining ReactElements as object literals` or something.